File: backend/src/models/Group.ts
import {
    Table,
    Column,
    CreatedAt,
    UpdatedAt,
    Model,
    DataType,
    PrimaryKey,
    AutoIncrement,
    HasMany,
    BelongsToMany,
    ForeignKey,
    BelongsTo
} from "sequelize-typescript";
import User from "./User";
import Role from "./Role";
import GroupRole from "./GroupRole";
import Tenant from "./Tenant";
import UserGroup from "./UserGroup";
import Permission from "./Permission";
import GroupPermission from "./GroupPermission";

@Table
class Group extends Model<Group> {
    @PrimaryKey
    @AutoIncrement
    @Column
    id: number;

    @Column(DataType.STRING)
    name: string;

    @ForeignKey(() => Tenant)
    @Column(DataType.UUID)
    tenantId: number | string;

    @BelongsTo(() => Tenant)
    tenant: Tenant;

    // Nova relação Many-to-Many: Um grupo pode ter múltiplos usuários
    @BelongsToMany(() => User, () => UserGroup)
    users: User[];

    @BelongsToMany(() => Role, () => GroupRole)
    roles: Role[];

    @BelongsToMany(() => Permission, () => GroupPermission)
    permissions: Permission[];

    @CreatedAt
    createdAt: Date;

    @UpdatedAt
    updatedAt: Date;
}

export default Group;

File: business/internal/models/group.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Group struct {
	ID        int       `gorm:"primaryKey" json:"id"`
	Name      string    `gorm:"not null" json:"name"`
	TenantID  uuid.UUID `gorm:"column:tenantId;type:uuid" json:"tenantId"`
	CreatedAt time.Time `gorm:"column:createdAt" json:"createdAt"`
	UpdatedAt time.Time `gorm:"column:updatedAt" json:"updatedAt"`

	// Relations
	Permissions []Permission `gorm:"many2many:group_permissions;joinForeignKey:groupId;joinReferences:permissionId" json:"permissions,omitempty"`
	Roles       []Role       `gorm:"many2many:group_roles;joinForeignKey:groupId;joinReferences:roleId" json:"roles,omitempty"`
}

func (Group) TableName() string {
	return "Groups"
}
